kinto/core/resource/__init__.py

```diff
@@ -460,17 +464,21 @@ def patch(self):
         existing = self._get_record_or_404(self.record_id)
         self._raise_412_if_modified(existing)
 
-        try:
-            # `data` attribute may not be present if only perms are patched.
-            changes = self.request.json.get('data', {})
-        except ValueError:
-            # If no `data` nor `permissions` is provided in patch, reject!
-            # XXX: This should happen in schema instead (c.f. ShareableViewSet)
-            error_details = {
-                'name': 'data',
-                'description': 'Provide at least one of data or permissions',
-            }
-            raise_invalid(self.request, **error_details)
+        # patch is specified as a list of of operations (RFC 6902)
+        if self._is_json_patch:
+            changes = {'data': self.request.json}
```

Comment:
Hmmm. Do I understand this correctly? We mutate `changes` because later we use `changes` to find any keys/values in the record that might have changed, which we do so that we know whether the record needs to be saved and also so we can respond with just changed fields.

If that's correct, then we're using `changes` as a strange bidirectional argument -- both passing data in, and getting data out. It seems to me like when it's input, `changes` is `{"data": [... list of operations ...]}`, but when it's output, it's `{"data": [... list of operations ...], "key1": "new_value1", "key2": "new_value2"}`. In other words, both kinds of data are mixed up. Is that true?

If I understand correctly, then I'd much rather have some other mechanism for returning the changed fields of the record. Perhaps `apply_changes` should return `(updated, changed_fields)`. Unfortunately this would be an API-breaking change. At the very least, we should document this behavior of `apply_changes` if we're going to take advantage of it, and I would say we should clear the old `data` entry when we update it.